egressgw: add support for excludedCIDRs

jibi opened the pull request:
Add support to the egress gateway manager for the new excludedCIDRs
policy property.

The implementation is based on the idea of adding, for each excluded
CIDR, "zero" entries to the egress policy map (i.e. entries with the
gateway IP set to 0.0.0.0). When traffic is then matched against these
entries in the datapath, it will skip altogether all the egress gateway
logic.

Fixes: https://github.com/cilium/cilium/issues/23002

jibi committed:
k8s: cegp: add excludedCIDRs field

ExcludedCIDRs can be used in a CiliumEgressGatewayPolicy to express a
list of destination CIDRs which will be excluded from the redirection
and SNAT logic.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: egressgw: handle excluded CIDRs case

Handle the case where an egress policy has a gateway IP set to 0:
any packet matching such policy should not go through egress gateway
redirection/SNAT.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: manager: cosmetic cleanup

In the callback we use in removeUnusedIpRulesAndRoutes() to match an IP
rule to an egress gateway policy, separate the preconditions from the
actual matching predicate.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
docs: egressgw: add section for excludedCIDRs

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann commented on bpf/lib/egress_policies.h:
```diff
@@ -46,6 +46,12 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
+	 */
+	if (!egress_gw_policy->gateway_ip)
+		return false;
+
```
Let's move this before the endpoint lookup :)

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml:
```diff
@@ -135,6 +135,14 @@ spec:
+              excludedCIDRs:
+                description: excludedCIDRs is a list of destination CIDRs which will
+                  be excluded from the egress gateway redirection and SNAT logic.
+                  It should be a subset of destinationCIDRs.
```
What happens if it isn't?

EDIT: Reading the code, it seems that nothing will happen if it isn't. Maybe we could clarify here that it will be ignored if it isn't a subset of `destinationCIDRs`?

jibi replied:
Correct, this was more like "to be effective it must be a subset" but nothing bad happens otherwise, let me add that :+1:

jibi committed:
bpf: tests: add unit test for egress gateway redirection

And rename existing test file to tc_egressgw_snat.c

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -246,11 +247,15 @@ func (manager *Manager) addMissingIpRulesAndRoutes(isRetry bool) (shouldRetry bo
 
+		if excludedCIDR {
+			return
+		}
```
Could you discuss the impact on those IP rules and routes in the commit description? In particular, based on these 3 lines, do they mean that some traffic that isn't matched by the policy based on exclusion CIDRs may still be matched by the IP rules?

```
$ cat ip-rule.md 
8:	from 10.20.0.67 to 35.241.162.0/24 lookup 302
```
So we may have the above even if we have an exclusion CIDR for 35.241.162.103? What's the impact of that? Why is it fine?

jibi replied:
> Why is it fine?

Nice catch, no, it's not fine as we need to exclude those CIDRs also from the IP rules/routes :disappointed: 

Need to think what's the best way to tackle this

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -385,8 +399,13 @@ func (manager *Manager) removeUnusedEgressRules() {
+		matchPolicy := func(endpointIP net.IP, dstCIDR *net.IPNet, excludedCIDR bool, gwc *gatewayConfig) bool {
+			gatewayIP := gwc.gatewayIP
+			if excludedCIDR {
+				gatewayIP = net.ParseIP("0.0.0.0")
```
```suggestion
				gatewayIP = zeroIPv4
```

pchaigno commented on pkg/egressgateway/policy.go:
```diff
@@ -159,32 +160,46 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig)
 			}
```
Thanks for defining this `isExcludedCIDR`! I miss Python for that.

jibi replied:
(or Rust :disappear:)

pchaigno commented on test/k8s/egress.go:
```diff
@@ -151,6 +150,27 @@ var _ = SkipDescribeIf(func() bool {
+			hostIP = k8s2IP
+		}
+
+		src, _ := fetchPodsWithOffset(kubectl, randomNamespace, "client", testDSClient, hostIP, false, 1)
```
Nit: Couldn't we merge this with `testEgressGateway` above with a flag to decide what case we're checking?

jibi replied:
Pushed https://github.com/cilium/cilium/pull/23448/commits/113ea0e834cfb59c7951cd276d3365d26eb3b677 to ease the review (and test it), will squash it with the manager changes once we are happy (tests are green https://jenkins.cilium.io/job/Cilium-PR-Tests-Kernel-Focus/522/)

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -305,9 +305,15 @@ nextIpRule:
+				return false
+			}
+
+			return ipRule.Src.IP.Equal(endpointIP) && ipRule.Dst.String() == dstCIDR.String()
```
Nice cleanup!

pchaigno reviewed (approved):
Nice pull request (as usual :smiley:)!

My only concern is with the ip rules and routes as I'm not sure I understand the impact there.

Regarding downgrade, should we ensure we get the commit into v1.13 before we merge this? Or alternatively, create a release blocker issue to ensure this PR doesn't end up in v1.14 before v1.13 is ready (in case something comes up and we forget to backport that commit)?
On the same topic, is this something we want in v1.13? (In which case we'll have to backport the commit to v1.12.)

jibi committed:
bpf: tests: add unit test for egress gateway excluded CIDRs

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: tests: use EGRESS_PREFIX_LEN macro

to initialize the egress gateway policies' lpm keys.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ip: export PartitionCIDR function

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/egressgateway/policy.go:
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
+	}
+
+	return cidrs
+}
```
Nit: This function alone would probably warrant its own commit.

I ran this on small examples and tried to turn it around. It looks good to me and I'm not seeing a more efficient implementation (unfortunately). We will have a lot of cases where we attempt to partition two CIDRs that don't overlap but luckily `PartitionCIDR` has special cases to handle that quickly.

That said, we probably don't want to call this monster too often.

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml:
```diff
@@ -135,6 +135,15 @@ spec:
                 - nodeSelector
                 type: object
+              excludedCIDRs:
+                description: excludedCIDRs is a list of destination CIDRs which will
```
```suggestion
                description: excludedCIDRs is a list of destination CIDRs that will
```

nathanjsweet commented on pkg/k8s/apis/cilium.io/v2/cegp_types.go:
```diff
@@ -52,6 +52,14 @@ type CiliumEgressGatewayPolicySpec struct {
 	// If a destination IP matches any one CIDR, it will be selected.
 	DestinationCIDRs []IPv4CIDR `json:"destinationCIDRs"`
 
+	// excludedCIDRs is a list of destination CIDRs which will be excluded
```
```suggestion
	// ExcludedCIDRs is a list of destination CIDRs that will be excluded
```

nathanjsweet reviewed (changes requested):
For API, a couple of nits.

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -27,7 +27,8 @@ import (
 )
 
```
> Moreover, to make IP rules and routes work with excluded CIDRs, ...

It would really help reviewing if the IP routes and rules changes would be in their own commit. Right now, this commit includes the basic support, the IP rules & routes handling, and all the helper functions. When reading the helper functions in particular, the lack of separation makes it even harder to understand why/how they are used.

pchaigno commented on pkg/egressgateway/policy.go:
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
+	}
+}
+
+func (config *PolicyConfig) forEachEndpointAndDestinationMinusExcludedCIDRs(epDataStore map[endpointID]*endpointMetadata,
```
What about `forEachEndpointAndDestination` becomes `forEachEndpointAndCIDR` while the new function becomes `forEachEndpointAndComputedDestination` (or just `forEachEndpointAndDestination` if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

In any case, you need comments on all of those helper functions to explain when it uses the input CIDR lists as is vs. the diff of the input CIDRs.

jibi replied:
> What about forEachEndpointAndDestination becomes forEachEndpointAndCIDR while the new function becomes forEachEndpointAndComputedDestination (or just forEachEndpointAndDestination if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L60
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L61

(so to me it reads like "iterate over all combinations of EP and destination CIDRs").

Same idea for the other `forEachEndpointAndDestinationMinusExcludedCIDRs` version, which although very mouthful, IMO makes it more clear what we are iterating over: EPs and `dstCIDRs` slice minus `excludedCDIRs` slice.

Perhaps for consistency I can rename the first one to `forEachEndpointAndDestinationCIDRs`

pchaigno replied:
> The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:

I think this is too complex to hope that the name can be self-explanatory. I don't think you can avoid adding a comment for those helper functions.

One issue with `forEachEndpointAndDestinationMinusExcludedCIDRs` is that it's unclear, without reading the function itself, if it iterates over all endpoints and destinations, but doesn't consider excluded CIDRs ("minus/except excluded CIDRs") or if it iterates over all endpoints and diffed destinations ("minus/diffed by excluded CIDRs").

jibi replied:
Fair enough, I'll add some comments then :+1:

jibi replied:
Scratch that, after taking another fresh look I'm also convinced that `forEachEndpointAndCIDR` and `forEachEndpointAndDestination` are better names, so I went for them :+1:

pchaigno replied:
Not to be a pain, but I think a quick comment would still help :slightly_smiling_face:

jibi replied:
Yes, I added them already in the latest push

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -27,7 +27,8 @@ import (
 )
```
> * forEachEndpointAndDestinationMinusExcludedCIDRs: works like
  forEachEndpointAndDestination, but iterate over
  destinationCIDRs - excludedCIDRs
> * matchesMinusExcludedCIDRs: works like matches but matches over
  destinationCIDRs - excludedCIDRs

Please replace `destinationCIDRs - excludedCIDRs` by actual words. Maybe I'm tired, but I didn't understand what you meant before I read the code. `but matches over the diff of destinationCIDRs minus excludedCIDRs` would probably be enough.

pchaigno commented on Documentation/network/egress-gateway.rst:
```diff
@@ -205,6 +205,20 @@ One or more IPv4 destination CIDRs can be specified with ``destinationCIDRs``:
+    - "a.b.c.0/24"
+
+In this case traffic destined to the ``a.b.0.0/16`` CIDR, except for the
+``a.b.c.0/24`` destination, will go through egress gateway and leave the cluster
```
Maybe we should make the two CIDRs completely unrelated in the example to not give the wrong impression that the excludedCIDR needs to be a subset of the destinationCIDR?

qmonnet replied:
Wrong impression? My understanding was the opposite?

```
              excludedCIDRs:
                description: ExcludedCIDRs is a list of destination CIDRs that will
                  be excluded from the egress gateway redirection and SNAT logic.
                  Should be a subset of destinationCIDRs otherwise it will not have
                  any effect.
```

jibi replied:
That's on purpose, excludedCIDRs needs indeed to be a subset of destinationCIDRs, otherwise it will have no effect

pchaigno replied:
I think I was confused on how CIDRs work :sweat_smile: If two CIDRs overlap, then they are either equal or one is the subset of the other.

pchaigno reviewed (approved):
Only minor comments from me, but the main commit (`egressgw: add support for excludedCIDRs`) could use some work to make it easier to follow. It's actually fairly simple changes if you already understand the main idea, but it's currently still hard to follow as is.

julianwiedmann commented on pkg/egressgateway/policy.go:
```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
+
+		for _, endpointIP := range endpoint.ips {
+			for _, cidr := range cidr {
+				if f(endpointIP, cidr, &config.gatewayConfig) {
```
I'm amazed by the `cidr := range cidr`. Can we give these two variables different names?

jibi replied:
I blame `:s/<something>/<something>/g` (the returned slice from `config.destinationMinusExcludedCIDRs()` should really be `cidrs`)

qmonnet commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -287,8 +403,15 @@ func newCiliumNode(name, nodeIP string, nodeLabels map[string]string) nodeTypes.
+		_, parsedExcludedCIDR, _ := net.ParseCIDR(excludedCIDR)
+		parsedExcludedCIDRs = append(parsedExcludedCIDRs, parsedExcludedCIDR)
+
+	}
```
Nit: blank line.

qmonnet reviewed (approved):
Looks good to me, thanks!

jibi committed:
egressgw: policy: add destinationMinusExcludedCIDRs helper

which returns the actual destination CIDRs space for a given policy,
which is defined by the destination CIDRs minus the excluded ones.

A practical example may be useful here. Let's say a policy specifies:

* destination CIDRs: 1.1.1.0/24
* excluded CIDRs: 1.1.1.1/2

The effective CIDR space for such policy will be the difference between
the two 1.1.1.0/24 and 1.1.1.1/24 CIDRs, so:

	1.1.1.0/28
	1.1.1.16/30
	1.1.1.20/31
	1.1.1.23/32
	1.1.1.24/29
	1.1.1.32/27
	1.1.1.64/26
	1.1.1.128/25

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: policy: parse excludedCIDRs from CEGP

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: policy: add excluded CIDRs iterating/matching helpers

This commit renames the forEachEndpointAndDestination method to
forEachEndpointAndCIDR, to make it more clear that it iterates over all
(destination and excluded) CIDRs of the receiver policy.

Next, it introduces 2 new helpers that allow iterating and matching over
all the endpoints and destination CIDRs minus the excluded ones:

* forEachEndpointAndDestinations: works like forEachEndpointAndCIDR, but
  iterates over destinationCIDRs minus excludedCIDRs
* matchesMinusExcludedCIDRs: works like matches but matches over
  destinationCIDRs minus excludedCIDRs

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: add support for excludedCIDRs

Add support to the egress gateway manager for the new excludedCIDRs
policy property.

The implementation is based on the idea of adding, for each excluded
CIDR, "zero" entries to the egress policy map (i.e. entries with the
gateway IP set to 0.0.0.0). When traffic is then matched against these
entries in the datapath, it will skip altogether all the egress gateway
logic.

Fixes: #23002
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann reviewed (approved):
:rocket: :rocket: :rocket:
